File: app/src/main/java/me/zhanghai/android/douya/ui/ImageLayout.java
/*
 * Copyright (c) 2015 Zhang Hai <Dreaming.in.Code.ZH@Gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.ui;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Build;
import androidx.annotation.RequiresApi;
import androidx.appcompat.widget.TintTypedArray;
import android.util.AttributeSet;
import android.widget.FrameLayout;
import android.widget.ImageView;

import butterknife.BindView;
import butterknife.ButterKnife;
import me.zhanghai.android.douya.R;
import me.zhanghai.android.douya.util.ImageUtils;
import me.zhanghai.android.douya.util.ViewUtils;

public class ImageLayout extends FrameLayout {

    public static final int FILL_ORIENTATION_HORIZONTAL = 0;
    public static final int FILL_ORIENTATION_VERTICAL = 1;

    @BindView(R.id.imagelayout_image)
    RatioImageView mImageView;
    @BindView(R.id.imagelayout_gif)
    ImageView mGifImage;

    public ImageLayout(Context context) {
        super(context);

        init(null, 0, 0);
    }

    public ImageLayout(Context context, AttributeSet attrs) {
        super(context, attrs);

        init(attrs, 0, 0);
    }

    public ImageLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        init(attrs, defStyleAttr, 0);
    }

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    public ImageLayout(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);

        init(attrs, defStyleAttr, defStyleRes);
    }

    @SuppressLint("RestrictedApi")
    private void init(AttributeSet attrs, int defStyleAttr, int defStyleRes) {

        setClickable(true);
        setFocusable(true);

        onInflateChildren();
        ButterKnife.bind(this);

        TintTypedArray a = TintTypedArray.obtainStyledAttributes(getContext(), attrs,
                R.styleable.ImageLayout, defStyleAttr, defStyleRes);
        int fillOrientation = a.getInt(R.styleable.ImageLayout_fillOrientation,
                FILL_ORIENTATION_HORIZONTAL);
        a.recycle();

        LayoutParams layoutParams = (LayoutParams) mImageView.getLayoutParams();
        layoutParams.width = fillOrientation == FILL_ORIENTATION_HORIZONTAL ?
                LayoutParams.MATCH_PARENT : LayoutParams.WRAP_CONTENT;
        layoutParams.height = fillOrientation == FILL_ORIENTATION_HORIZONTAL ?
                LayoutParams.WRAP_CONTENT : LayoutParams.MATCH_PARENT;
        mImageView.setLayoutParams(layoutParams);
    }

    protected void onInflateChildren() {
        ViewUtils.inflateInto(R.layout.image_layout, this);
    }

    public void loadImage(SizedImageItem image) {
        ImageUtils.loadImageWithRatio(mImageView, image);
        ViewUtils.setVisibleOrGone(mGifImage, image.isAnimated());
    }

    public void releaseImage() {
        mImageView.setImageDrawable(null);
    }
}

File: app/src/main/java/me/zhanghai/android/douya/ui/ImageLayout.kt
/*
 * Copyright (c) 2020 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.ui

import android.content.Context
import android.util.AttributeSet
import android.widget.FrameLayout
import androidx.core.view.doOnAttach
import androidx.core.view.updateLayoutParams
import androidx.lifecycle.MutableLiveData
import me.zhanghai.android.douya.api.info.SizedImage
import me.zhanghai.android.douya.api.util.normalOrClosest
import me.zhanghai.android.douya.arch.ResumedLifecycleOwner
import me.zhanghai.android.douya.arch.mapDistinct
import me.zhanghai.android.douya.databinding.ImageLayoutBinding
import me.zhanghai.android.douya.util.layoutInflater

class ImageLayout : FrameLayout {
    private val lifecycleOwner = ResumedLifecycleOwner()

    private val binding = ImageLayoutBinding.inflate(context.layoutInflater, this, true)

    val viewModel = ViewModel()

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int
    ) : super(context, attrs, defStyleAttr)

    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        doOnAttach {
            val layoutParams = layoutParams
            binding.imageImage.updateLayoutParams {
                width = layoutParams.width
                height = layoutParams.height
            }
        }

        binding.lifecycleOwner = lifecycleOwner
        binding.viewModel = viewModel
    }

    fun setImage(image: SizedImage?) {
        viewModel.setImage(image)
        binding.executePendingBindings()
    }

    class ViewModel {
        data class State(
            val ratio: Float,
            val url: String,
            val isGif: Boolean
        )

        private val state = MutableLiveData(
            State(
                ratio = 1f,
                url = "",
                isGif = false
            )
        )
        val ratio = state.mapDistinct { it.ratio }
        val url = state.mapDistinct { it.url }
        val isGif = state.mapDistinct { it.isGif }

        fun setImage(image: SizedImage?) {
            val imageItem = image?.normalOrClosest
            state.value = State(
                ratio = imageItem?.let { it.width.toFloat() / it.height } ?: 1f,
                url = imageItem?.url ?: "",
                isGif = image?.isAnimated ?: false
            )
        }

        fun open() {
            // TODO
        }
    }
}
